File: java/var-core/src/test/java/dev/varar/core/StepRoleTest.java
package com.oselvar.var.core;

import static org.junit.jupiter.api.Assertions.assertEquals;

import java.util.List;
import org.junit.jupiter.api.Test;

/** Translated from {@code var-core/tests/step-role.test.ts}. */
class StepRoleTest {

    @Test
    void noStepAfterTheSelectionMeansSensorExpectationLast() {
        var neighbours = new StepRole.Neighbours(List.of(StepKind.STIMULUS), List.of());
        assertEquals(StepKind.SENSOR, StepRole.inferStepRole(neighbours));
    }

    @Test
    void aSensorFollowsAndNoActionSitsBetweenMeansAction() {
        var neighbours = new StepRole.Neighbours(List.of(StepKind.STIMULUS), List.of(StepKind.SENSOR));
        assertEquals(StepKind.STIMULUS, StepRole.inferStepRole(neighbours));
    }

    @Test
    void nothingBeforeAndAStepAfterMeansContext() {
        var neighbours = new StepRole.Neighbours(List.of(), List.of(StepKind.STIMULUS));
        assertEquals(StepKind.STIMULUS, StepRole.inferStepRole(neighbours));
    }

    @Test
    void otherwiseMeansAction() {
        var neighbours = new StepRole.Neighbours(List.of(StepKind.STIMULUS), List.of(StepKind.STIMULUS));
        assertEquals(StepKind.STIMULUS, StepRole.inferStepRole(neighbours));
    }
}

File: rust/core/tests/step_role_test.rs
//! Port of `StepRoleTest.java` / `step-role.test.ts`.

use varar_core::step_kind::StepKind;
use varar_core::step_role::{Neighbours, infer_step_role};

#[test]
fn no_step_after_the_selection_means_sensor_expectation_last() {
    let neighbours = Neighbours::new(vec![StepKind::Stimulus], vec![]);
    assert_eq!(StepKind::Sensor, infer_step_role(&neighbours));
}

#[test]
fn a_sensor_follows_and_no_action_sits_between_means_action() {
    let neighbours = Neighbours::new(vec![StepKind::Stimulus], vec![StepKind::Sensor]);
    assert_eq!(StepKind::Stimulus, infer_step_role(&neighbours));
}

#[test]
fn nothing_before_and_a_step_after_means_context() {
    let neighbours = Neighbours::new(vec![], vec![StepKind::Stimulus]);
    assert_eq!(StepKind::Stimulus, infer_step_role(&neighbours));
}

#[test]
fn otherwise_means_action() {
    let neighbours = Neighbours::new(vec![StepKind::Stimulus], vec![StepKind::Stimulus]);
    assert_eq!(StepKind::Stimulus, infer_step_role(&neighbours));
}
